packages/utils/src/namespaces.ts

```diff
@@ -63,3 +64,143 @@ export function getRequiredNamespacesFromNamespaces(
+
+    namespaces[namespace] = {
+      chains: supportedChains,
+      methods: supportedMethods,
+      events: supportedEvents,
+      accounts: supportedAccounts,
+    };
+  });
+
+  // verify all required namespaces are supported
+  const err = isConformingNamespaces(requiredNamespaces, namespaces, "approve()");
+  if (err) throw new Error(err.message);
+
+  const approvedNamespaces = {};
+
+  // assign accounts for the required namespaces
+  Object.keys(normalizedRequired).forEach((requiredNamespace) => {
+    const chains = supportedNamespaces[requiredNamespace].chains.filter((chain) =>
+      normalizedRequired[requiredNamespace]?.chains?.includes(chain),
+    );
+    const methods = supportedNamespaces[requiredNamespace].methods.filter((method) =>
+      normalizedRequired[requiredNamespace]?.methods?.includes(method),
+    );
+    const events = supportedNamespaces[requiredNamespace].events.filter((event) =>
+      normalizedRequired[requiredNamespace]?.events?.includes(event),
+    );
+
+    approvedNamespaces[requiredNamespace] = {
+      chains,
+      methods,
+      events,
+      accounts: chains
+        .map((chain: string) =>
+          supportedNamespaces[requiredNamespace].accounts.filter((account: string) =>
+            account.includes(chain),
+          ),
+        )
+        .flat(),
+    };
+  });
+
+  // add optional namespaces
+  Object.keys(normalizedOptional).forEach((optionalNamespace) => {
+    if (!supportedNamespaces[optionalNamespace]) return;
+
+    const chainsToAdd = normalizedOptional[optionalNamespace]?.chains?.filter((chain) =>
+      supportedNamespaces[optionalNamespace].chains.includes(chain),
+    );
+    const methodsToAdd = supportedNamespaces[optionalNamespace].methods.filter((method) =>
+      normalizedOptional[optionalNamespace]?.methods?.includes(method),
+    );
+    const eventsToAdd = supportedNamespaces[optionalNamespace].events.filter((event) =>
+      normalizedOptional[optionalNamespace]?.events?.includes(event),
+    );
+
+    const accountsToAdd = chainsToAdd
+      ?.map((chain: string) =>
+        supportedNamespaces[optionalNamespace].accounts.filter((account: string) =>
+          account.includes(chain),
+        ),
+      )
+      .flat();
+
+    approvedNamespaces[optionalNamespace] = {
+      chains: mergeArrays(approvedNamespaces[optionalNamespace]?.chains, chainsToAdd),
+      methods: mergeArrays(approvedNamespaces[optionalNamespace]?.methods, methodsToAdd),
+      events: mergeArrays(approvedNamespaces[optionalNamespace]?.events, eventsToAdd),
+      accounts: mergeArrays(approvedNamespaces[optionalNamespace]?.accounts, accountsToAdd),
+    };
+  });
+
+  return approvedNamespaces;
+}
+
+export function isCaipNamespace(namespace: string): boolean {
+  return namespace.includes(":");
+}
+
+export function parseNamespaceKey(namespace: string) {
+  return isCaipNamespace(namespace) ? namespace.split(":")[0] : namespace;
+}
+
+/**
+ * Converts
+ * ```
```

Comment:
Took the liberty of pushing a tiny change for these backticks so editors pick up the code blocks correctly:

<img width="743" alt="Screenshot 2023-03-29 at 21 21 30" src="https://user-images.githubusercontent.com/8663099/228777051-6431fe60-cf8c-4aac-ada9-514361a1ac77.png">